ENH: stats: Mixture distribution

thomaskeck opened the pull request:
A mixture distribution generator "mixture_gen" to create a superposition of several other scipy distributions. This can be used to combine several components with fixed relative fractions into one pdf, or to create for instance a triple-gaussian distribution (which is used to model gaussian noise coming from different sources).

statsmodels.distributions may provide similar functionality, as was pointed out by @josef-pkt 

This is a split of, of another PR which contained several other distributions #6795

thomaskeck commented:
Fixed conflicts.
Await further feedback.

@ev-br 
The online conflict solution editor merged the master branch, in #6801 you comment
that you don't recommend doing so.
So, should I avoid using the online editor in the future for scipy branches?

ev-br commented:
Yeah, the preferred way seems to be the same as before: rebase locally and force-push.

thomaskeck commented:
Ping.

Are there further comments or suggestions to this PR?

thomaskeck commented:
Hello, this is my bi-monthly reminder for this PR.
I understand that the PR is probably forgotten and not shown anymore on the first page of open PRs.
But still I try to get some attention by this comment :-)

There are conflicts, which I will resolve if there is a chance that this PR will be merged.

@ev-br Since you are provided me feedback in the past I mention you directly, which I should have done maybe earlier :-)

perimosocordiae commented on scipy/stats/_continuous_distns.py:
```diff
@@ -5392,8 +5392,165 @@ def _updated_ctor_param(self):
+    """
+    Returns all shape parameters of a scipy distribution
+    including loc and scale
+    @param distribution scipy.stats.rv_continuous
```
Use the numpydoc style.

thomaskeck replied:
I will change this (probably next weekend, I'm currently on vacation). Thanks for the feedback.

perimosocordiae commented on scipy/stats/_continuous_distns.py:
```diff
@@ -5392,8 +5392,165 @@ def _updated_ctor_param(self):
+    shape_parameters = []
+    if distribution.shapes is not None:
+        shape_parameters += distribution.shapes.split(', ')
+    shape_parameters += ['loc', 'scale']
```
Why not initialize with `['loc', 'scale']`?

thomaskeck replied:
If I remeber correctly the reason is that loc and scale must be the last parameters because at least at some points in the scipy.stats code the arguments to a distirbution are passed as args (not as kwargs), so the order matters, and the order is "shape_parameters, loc, scale".

thomaskeck replied:
Maybe to given an example, in line 5512 of the same file I call the pdf method of a distribution like this. 
 distribution.pdf(x, *shape)

The last two elements of the shape variable must be loc and scale.

thomaskeck commented:
Updated version of this PR:

-  resolved the conflicts
-  migrated the tests to pytest
-  updated the docstring of _get_shape_parameters to numpydoc style
-  adapted  and improved the code in test_distribution (exception is now re-raised if necessary)

Note for people with a (high energy) physics background:
I also want to mention here that I have some custom code which builds on scipy.stats and rv_mixture
and implements a ML (maximum likelihood) fitting framework like it is used in many physics analyses to extract observables. So far physicist used e.g. RooFit to do this.
The code is (probably) to specific  for scipy, but as soon as rv_mixture is part of scipy I will polish the code and push it on github.

thomaskeck commented:
@ev-br @perimosocordiae Any more suggestions for this PR?

pvanmulbregt commented:
A couple of questions:
1. Does this work when given frozen distributions?  Thinking about how I would use it, I'd be likely to create a number of frozen distributions, probably of the same type, and want to feed those to the mixture.  Is that supported ?  E.g.
```
means = np.array([1, 2, 3])         # example means/variance for a mixture of three 1-dim Gaussians
vars = np.array([1.0, 4.0, 25.0])
weights = [1, 4, 1]

# Create the mixture
dists = [scipy.stats.norm(loc=m, scale=np.sqrt(v)) for m,v in zip(means, vars)]
mixture = scipy.stats.rv_mixture(dists, weights)

#Ccompute the pdf for x=2.0...
mixture.pdf(2.0)
```
or would I have to do something like 
```
# Create the mixture
iargs = [('Gauss_%d' % k, scipy.stats.norm) for k in range(len(means))]
mixture = scipy.stats.rv_mixture(* iargs)

# Compute the pdf for x=2.0...
d = dict()
for k in range(len(means)):
    d['Gauss_%d_norm' % k] = weights[k]
    d['Gauss_%d_loc' % k] = means[k]
    d['Gauss_%d_scale' % k] = np.sqrt(vars[k])
# So d is {'Gauss_0_norm': 1, 'Gauss_0_loc': 1, 'Gauss_0_scale': 1.0, 'Gauss_1_norm': 4, 'Gauss_1_loc': 2, 'Gauss_1_scale': 2.0, 'Gauss_2_norm': 1, 'Gauss_2_loc': 3, 'Gauss_2_scale': 5.0}
mixture.pdf(2.0, **d)
```

2. Why the suffix `"_norm"` for the weight on the component distributions?  The co-occurrence of `"_norm"` and `"norm"` (in `scipy.stats.norm`) is a little confusing, as "norm" isn't a word but is short for "normalization"(?).  Personally I would have found "weight" say, easier to understand what it was doing.
